### PR TITLE
Feature/translate into typescript

### PR DESCRIPTION
**Pull request by sihu**

### CODE REVIEW

File: src/renuo_upload.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      //TODO check if necessary, unexpected behavior, good point from Y
+      jQuery.each(result, (k, v) => {
+        const parentForm:JQuery = jQuery(this.element).parents('form');
+        parentForm.append(`<input type='hidden' name='renuoupload[${name}][${k}]' value='${v}'>`);
+      });
+    }
+  }
+}
+
+window['RenuoUpload'] = RenuoUpload;
```
Comment (lukaselmer) [inline src/renuo_upload.ts]: window.RenuoUpload = doesn't work (in TypeScript), right?

File: tslint.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{
```
Comment (lukaselmer) [inline tslint.json:1]: Nice!

File: typings.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "ambientDependencies": {
+    "dropzone": "registry:dt/dropzone#4.0.1+20160316155526",
```
Comment (lukaselmer) [inline typings.json]: Latest version?

File: src/renuo_upload.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    if (typeof this.callback !== 'function')
+      this.callback = this.defaultCallback;
+  }
+
+  private defaultCallback(result:RenuoUploadResult) {
+    if (jQuery(this.element).parents('form').length) {
+      name = result.name;
+      //TODO discuss with Lukas about upload of two times the same file
+      delete result.name;
+      //TODO check if necessary, unexpected behavior, good point from Y
```
Comment (lukaselmer) [inline src/renuo_upload.ts]: We should resolve these todos, right @cyrilkyburz ?

File: gulpfile.js [hunk 5]
```diff
@@ -1,2 +1,31 @@
...
+  gulp.src( paths.src)
+    .pipe(typescript())
+    .pipe(uglify())
+    .pipe(rename({extname: '.min.js'}))
+    .pipe(gulp.dest(paths.dist))
+});
+
+gulp.task('watch', function(){
+  gulp.watch([paths.src], ['compile']);
+});
```
Comment (lukaselmer) [inline gulpfile.js:105]: Does compile call typescript-compile?

File: package.json [hunk 6]
```diff
@@ -13,20 +13,22 @@
...
-    "gulp-coffee": "^2.3.1",
+    "gulp": "^3.9.1",
     "gulp-rename": "^1.2.0",
-    "gulp-uglify": "^1.1.0",
+    "gulp-tsc": "^1.1.5",
+    "gulp-uglify": "^1.5.3",
     "gulp-util": "^2.2.14",
-    "jasmine-core": "^2.2.0",
-    "jasmine-expect": "^2.0.0-beta1",
+    "jasmine-core": "^2.2.20",
```
Comment (lukaselmer) [inline package.json]: Are there any tests at the moment? If not i would remove the dependency.

**Comment by lukaselmer** [comment]
Imo it lgtm. What do you think @cyrilkyburz ?

File: typings.json (hunk 3 above)
Comment (sihu) [inline typings.json]: latest released version! still waiting for release 4.0.3 to be merged

File: src/renuo_upload.ts (hunk 4 above)
Comment (cyrilkyburz) [inline src/renuo_upload.ts]: Yes, would be a good moment to solve them.

Problem 1:  Second file uploaded with the same name will override the first
Problem 2: The delete is fishy (we remove the name attribute from the object and iterate over all other attributes)

Problem 1: I think the behaviour is okay, but we have always a "dead" file on the server. Do we need to take action?
Problem 2: We could rewrite it with an if, would probably be more readable, or any other suggestion?

File: package.json [hunk 7]
```diff
@@ -13,20 +13,22 @@
   "bugs": {
     "url": "https://redmine.renuo.ch/projects/upload/issues"
   },
-  "dependencies": {},
+  "dependencies": {
+    "bower": "^1.7.9"
```
Comment (cyrilkyburz) [inline package.json]: bower is just a dev dependency..

File: package.json (hunk 6 above)
Comment (cyrilkyburz) [inline package.json]: there are no test :(

File: src/renuo_upload.ts (hunk 1 above)
Comment (cyrilkyburz) [inline src/renuo_upload.ts]: Sure, I always do it this way. Also dropzone: https://github.com/enyo/dropzone/blob/master/src/dropzone.coffee#L1467

File: src/renuo_upload.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  private initializeDropzone() {
+    Dropzone.autoDiscover = false;
+
+    jQuery(this.element).addClass('dropzone');
+    const uploadDropzone:Dropzone = new Dropzone(this.element, this.dropzoneOptions);
+
+    uploadDropzone.on('success', (file) => this.callback(this.buildResult(file)));
+
+    uploadDropzone.on('error', () => {
+      //todo inform sentry/new relic
```
Comment (cyrilkyburz) [inline src/renuo_upload.ts]: This feature would be really nice... don't know if you have time?

**Comment by cyrilkyburz** [comment]
Do we want to provide typings for the renuo-upload? (Now when it is typescript :D)

Like: https://github.com/renuo/nievergelt-frontend/blob/develop/app/typings/renuo_upload.d.ts

File: src/renuo_upload.ts (hunk 1 above)
Comment (sihu) [inline src/renuo_upload.ts]: typescript is not happy about this syntax @lukaselmer, so this way it works!

File: typings.json (hunk 3 above)
Comment (sihu) [inline typings.json]: has been merged now! so i can change it!

File: src/renuo_upload.ts [hunk 9]
```diff
@@ -121,12 +121,10 @@ class RenuoUpload {
...
 
   private defaultCallback(result:RenuoUploadResult) {
+    // TODO discuss with Lukas about upload of two times the same file
     if (jQuery(this.element).parents('form').length) {
-      name = result.name;
-      //TODO discuss with Lukas about upload of two times the same file
-      delete result.name;
-      //TODO check if necessary, unexpected behavior, good point from Y
       jQuery.each(result, (k, v) => {
+        if(k == 'name') return true;
```
Comment (cyrilkyburz) [inline src/renuo_upload.ts]: missing space 🐼

File: typings.json (hunk 3 above)
Comment (lukaselmer) [inline typings.json]: great!

File: bower.json [hunk 10]
```diff
@@ -7,7 +7,9 @@
   "private": false,
   "dependencies": {
     "jquery": "latest",
-    "dropzone": "git://github.com/cyrilkyburz/dropzone"
+    "dropzone": "git://github.com/cyrilkyburz/dropzone",
+    "jasmine-core": "jasmine#^2.4.1",
```
Comment (cyrilkyburz) [inline bower.json]: wrong dependencies,we can now use official dropzone & we need jasmine only in development

Comment (lukaselmer) [inline bower.json]: 👍

**Comment by lukaselmer** [comment]
> Do we want to provide typings for the renuo-upload? (Now when it is typescript :D)
> Like: https://github.com/renuo/nievergelt-frontend/blob/develop/app/typings/renuo_upload.d.ts

just set the `decalaration` option to true 👍 
https://github.com/ivogabe/gulp-typescript#options

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    - $(npm config get prefix)/bin/tslint
+    - $(npm config get prefix)/bin/gulp
+before_install:
+  - export TZ=Europe/Zurich
+before_script:
+  - export DISPLAY=:99.0
+  - sh -e /etc/init.d/xvfb start
+  - echo {\"token\":\"$TSDRC_TOKEN\"} > .tsdrc
+  - bin/setup
+script: bin/check
```
Comment (lukaselmer) [inline .travis.yml:26]: can we use bin/setup instead of bin/check here? bin/setup calls bin/check in the end, right?

File: .nvmrc [hunk 12]
```diff
@@ -0,0 +1 @@
+5
```
Comment (lukaselmer) [inline .nvmrc:1]: 👍

File: bin/setup [hunk 13]
```diff
@@ -0,0 +1,8 @@
+nvm install
+npm install -g npm
+npm install -g bower gulp coffee-script typings
+typings install
+npm install
+
+grep '127.0.0.1 renuo-upload.dev' /etc/hosts || echo '127.0.0.1 renuo-upload.dev' | sudo tee -a /etc/hosts
+cat .git/hooks/pre-commit > /dev/null || ln -s ../../bin/check .git/hooks/pre-commit
```
Comment (lukaselmer) [inline bin/setup:8]: 😮 nope, we don't

we should call bin/check here

maybe use the latest version of the setup guide here?

File: bower.json [hunk 14]
```diff
@@ -7,10 +7,12 @@
   "private": false,
   "dependencies": {
     "jquery": "latest",
-    "dropzone": "git://github.com/cyrilkyburz/dropzone"
+    "dropzone": "^4.3.0"
```
Comment (lukaselmer) [inline bower.json:10]: yay!

File: bower.json [hunk 15]
```diff
@@ -7,10 +7,12 @@
...
   },
   "devDependencies": {
-    "renuo-upload": "https://github.com/renuo/renuo-upload"
+    "renuo-upload": "https://github.com/renuo/renuo-upload",
```
Comment (lukaselmer) [inline bower.json]: mm? this looks like inception?

File: demo/index.html [hunk 16]
```diff
@@ -0,0 +1,57 @@
+<!DOCTYPE html>
```
Comment (lukaselmer) [inline demo/index.html:1]: very nice!

File: demo/index.html [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+  $(document).ready(function () {
+    var dropzoneOptions = {
+      acceptedFiles: 'image/*,application/pdf',
+      dictDefaultMessage: 'Ziehen Sie eine Bild-Datei oder ein PDF hier herein, oder klicken Sie hier um die Datei hochzuladen.',
+      dictFallbackMessage: 'Sie verwenden einen veralteten Browser was das verwenden des Uploads verhindert.',
+      dictInvalidFileType: 'Hier können nur Bild-Dateien hochgeladen werden.',
+      dictFileTooBig: 'Die Datei ist zu gross.'
+    };
+    new window.RenuoUpload($('.renuo-upload'), dropzoneOptions, function (a, b, c) {
+      console.log(a);
```
Comment (lukaselmer) [inline demo/index.html]: like @cyrilkyburz already noticed: we could use some more sophisticated logging here (see readme). agree?

File: karma.conf.coffee [hunk 18]
```diff
@@ -0,0 +1,89 @@
+'use strict'
```
Comment (lukaselmer) [inline karma.conf.coffee:1]: 👍

File: src/renuo_signing_response.d.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
+interface RenuoSigningResponse {
```
Comment (lukaselmer) [inline src/renuo_signing_response.d.ts]: why do we declare it as d.ts? should one always do that for interfaces?

File: src/renuo_upload.d.ts [hunk 20]
```diff
@@ -0,0 +1,4 @@
+declare class RenuoUpload {
```
Comment (lukaselmer) [inline src/renuo_upload.d.ts]: why declare? why .d.ts?

Comment (lukaselmer) [inline src/renuo_upload.d.ts]: imo this should be automatically generated and not be part of the src

File: src/renuo_upload.spec.ts [hunk 21]
```diff
@@ -0,0 +1,11 @@
+///<reference path="../typings/browser/ambient/jasmine/index.d.ts"/>
+///<reference path="renuo_upload.ts"/>
+
+describe('ContentBlockDrawer', function () {
+  const upload = RenuoUpload;
+
+  it('draws simple text', function () {
+    expect(upload).toBeTruthy();
+    expect(true).toBeTruthy();
```
Comment (lukaselmer) [inline src/renuo_upload.spec.ts:9]: i guess we should remove the second spec

File: src/renuo_upload.ts [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+    if (!this.dropzoneOptions.renameFilename) this.dropzoneOptions.renameFilename = this.cleanFilename;
+  }
+
+  private adaptCallback() {
+    if (typeof this.callback !== 'function')
+      this.callback = this.defaultCallback;
+  }
+
+  private defaultCallback(result:RenuoUploadResult) {
+    // TODO discuss with Lukas about upload of two times the same file
```
Comment (lukaselmer) [inline src/renuo_upload.ts]: change the todo to: fix the problem with duplicated image names

File: src/renuo_upload_result.d.ts [hunk 23]
```diff
@@ -0,0 +1,9 @@
+interface RenuoUploadResult {
```
Comment (lukaselmer) [inline src/renuo_upload_result.d.ts]: d.ts?

File: tsconfig.json [hunk 24]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "noImplicitAny": true,
+    "sourceMap": false,
+    "outFile": "renuo-upload.js",
+    "inlineSourceMap": true
```
Comment (lukaselmer) [inline tsconfig.json:8]: add declaration: true ?

File: typings/browser.d.ts [hunk 25]
```diff
@@ -0,0 +1,5 @@
+/// <reference path="browser/ambient/dropzone/index.d.ts" />
```
Comment (lukaselmer) [inline typings/browser.d.ts:1]: if we commit these typings, we probably should also commit the other typings?

**Comment by lukaselmer** [comment]
imo some small corrections are necessary, but in general it lgtm 👍

**Pull request merged**